runner.py

```diff
@@ -1,138 +1,242 @@
+        HostPCGenericTarget,
+        IA32GenericQemuTarget,
+        RISCV64GenericQemuTarget,
+        ARMv7A9Zynq7000QemuTarget,
+        STM32L4x6Target,
+        IMXRT106xEvkTarget,
+        IMXRT117xEvkTarget,
+        Zynq7000ZedboardTarget,
+    ]
+
+    hosts = [EmulatorHost, RpiHost]
+
+    try:
+        ext_targets, ext_hosts = load_extensions()
+    except ExtensionError as e:
+        print(e)
+        sys.exit(4)
+
+    targets.extend(ext_targets)
+    hosts.extend(ext_hosts)
+
+    # NOTE default targets/hosts can be overwritten if extensions have the same name!
+    targets = {t.name: t for t in targets}
+    hosts = {h.name: h for h in hosts}
+
+    return targets, hosts
 
 
 def main():
-    args = parse_args()
-    set_logger(args.log_level)
-
-    runner = TestsRunner(target=args.target,
-                         test_paths=args.test,
-                         build=args.build,
-                         flash=not args.no_flash,
-                         serial=(args.serial, args.baudrate),
-                         log=(args.log_level == logging.DEBUG),
-                         long_test=args.long_test
-                         )
-
-    passed, failed, skipped = runner.run()
-
-    total = passed + failed + skipped
-    summary = f'TESTS: {total}'
-    summary += f' {Color.colorify("PASSED", Color.OK)}: {passed}'
-    summary += f' {Color.colorify("FAILED", Color.FAIL)}: {failed}'
-    summary += f' {Color.colorify("SKIPPED", Color.SKIP)}: {skipped}\n'
-    logging.info(summary)
-
-    if failed == 0:
-        print("Succeeded!")
-        sys.exit(0)
-    else:
-        print("Failed!")
-        sys.exit(1)
+    targets, hosts = resolve_targets_and_hosts()
+
+    args = parse_args(targets, hosts)
+    host_cls = hosts[args.host]
+    host = host_cls()
+
+    # Create partial context for target initialization
+    ctx = TestContext(
+        target=None,
+        host=host,
+        port=args.port,
+        baudrate=args.baudrate,
+        project_path=resolve_project_path(),
+        nightly=args.nightly,
+        should_flash=not args.no_flash,
+        should_test=not args.no_test,
+        verbosity=args.verbose,
+    )
+
+    try:
+        target_cls = targets[args.target]
+        target = target_cls.from_context(ctx)
+    except PortError as e:
+        # TODO Make port finding in the global scope
+        print(e)
+        sys.exit(2)
+
+    ctx = TestContext(
+        target=target,
+        host=ctx.host,
+        port=ctx.port,
+        baudrate=ctx.baudrate,
+        project_path=ctx.project_path,
+        nightly=ctx.nightly,
+        should_flash=ctx.should_flash,
+        should_test=ctx.should_test,
+        verbosity=ctx.verbosity,
+    )
+
+    # Set global context
+    trunner.ctx = ctx
+
+    runner = TestRunner(
+        ctx=ctx,
+        test_paths=args.test,
+    )
+
+    ok = runner.run()
+    if not ok:
+        return 1
+
+    return 0
 
 
 if __name__ == "__main__":
-    main()
+    exit(main())
```

Comment:
`exit` is meant for python cli, `sys.exit` should be used here. Though I've never seen exit(main()) in python code before so I believe it's not idiomatic, also there is sys.exit(2) in LOC 212 you should either return there 2 of perform only sys.exit on failures in main

Reply:
changed to sys.exit in case of error